src/hotspot/share/runtime/os.cpp

```diff
@@ -2105,15 +2105,22 @@ void os::pretouch_memory(void* start, void* end, size_t page_size) {
     // We're doing concurrent-safe touch and memory state has page
     // granularity, so we can touch anywhere in a page.  Touch at the
     // beginning of each page to simplify iteration.
-    char* cur = static_cast<char*>(align_down(start, page_size));
+    void* first = align_down(start, page_size);
     void* last = align_down(static_cast<char*>(end) - 1, page_size);
-    assert(cur <= last, "invariant");
+    assert(first <= last, "invariant");
     // Iterate from first page through last (inclusive), being careful to
     // avoid overflow if the last page abuts the end of the address range.
-    for ( ; true; cur += page_size) {
-      Atomic::add(reinterpret_cast<int*>(cur), 0, memory_order_relaxed);
-      if (cur >= last) break;
-    }
+    pd_pretouch_memory(first, last, page_size);
+  }
+}
+
+void os::pretouch_memory_common(void* first, void* last, size_t page_size) {
+  assert(is_aligned(first, page_size), "pointer " PTR_FORMAT " is not page-aligned by %zu", p2i(first), page_size);
+  assert(is_aligned(last, page_size), "pointer " PTR_FORMAT " is not page-aligned by %zu", p2i(last), page_size);
```

Comment:
New assertions. Affects all platforms, not only Linux. If possible, please don't change behavior on other platforms.

Have you checked all users of this function, also on other platforms? AFAICT calling this with unaligned pointers would have been possible and would have worked.

Reply:
The assertions were suggested by Kim Barrett [here](https://github.com/openjdk/jdk/pull/15781#discussion_r1340872840). The function is private in os, and supposed to be called by pd_pretouch_memory. I'm fine without the assertions personally.

Reply:
Okay, in that case its fine.